client/scala/src/main/scala/ai/verta/blobs/dataset/S3.scala

```diff
@@ -20,13 +22,61 @@ import scala.annotation.tailrec
  *  val s3Blob2: Try[S3] = S3(S3Location("some-path"))
  *  }}}
  */
-case class S3(protected val contents: HashMap[String, FileMetadata]) extends Dataset {
+case class S3(
+  protected val contents: HashMap[String, FileMetadata],
+  protected val enableMDBVersioning: Boolean = false
+) extends Dataset {
   /** Get the version id of a file
    *  @param path: S3 URL of a file in the form "s3://<bucketName>/<key>"
    *  @return the version id of the file, if the file exists and has a versionId; otherwise None.
    */
   def getVersionId(path: String) = contents.get(path).flatMap(_.versionId)
 
+  /** Preparing for uploading by downloading the stored objects from S3.
+   *  @return A map from the s3 path to the path of the stored object, if succeeds
+   */
+  override def prepareForUpload(): Try[Map[String, UploadComponent]] = {
+    Try(contents
+      .mapValues(metadata =>
+        toUploadComponent(S3Location(metadata.path, metadata.versionId).get, S3.s3, metadata).get
+      )
+      .toMap
+    )
+  }
+
+  /** Helper method to download the object stored in a S3 location and construct the corresponding upload component
+   *  Based on https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/examples-s3-objects.html#download-object
+   *  @param location the S3 location. Must correspond to a file
+   *  @param s3 The s3 client
+   *  @return the path of the saved object, computed based on the s3 location, if succeeds.
+   */
+  private def toUploadComponent(location: S3Location, s3: AmazonS3, metadata: FileMetadata) = Try {
+    val file = File.createTempFile(location.key.get, null)
+
+    val request =
+      if(location.versionID.isDefined)
+        new GetObjectRequest(location.bucketName, location.key.get, location.versionID.get)
+      else
+        new GetObjectRequest(location.bucketName, location.key.get)
+
+    val obj: S3Object = s3.getObject(request)
+    val s3InputStream: S3ObjectInputStream = obj.getObjectContent()
+    val fileOutputStream: FileOutputStream = new FileOutputStream(file)
+    val buffer = new Array[Byte](1024)
+    var readLen = s3InputStream.read(buffer)
+
+    while (readLen > 0) {
+        fileOutputStream.write(buffer, 0, readLen)
+        readLen = s3InputStream.read(buffer)
+    }
```

Comment:
Isn't there a function to do this for you automatically? I bet there's some optimized thing to copy form input to output stream.

Reply:
No idea lol. I literally copy pasted this section from Amazon S3 documentation =) I'll figure sth out.

Reply:
There is, but it only exists for Java 9 I think:
https://stackoverflow.com/questions/43157/easy-way-to-write-contents-of-a-java-inputstream-to-an-outputstream
Other answers mostly involve some third party libraries.

Reply:
Got it.